### PR TITLE
[IOTDB-5]Combine deletion with FileNodeProcessor and FileNodeManager

### PR DESCRIPTION
**Pull request by jt2594838**
Implemented deletion-related interfaces in FileNodeProcessor and FileNodeManager.

### CODE REVIEW

File: iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
+
+/**
+ * SimpleFileVersionController uses a local file and its file name to store the version.
+ */
+public class SimpleFileVersionController implements VersionController {
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java]: Suppose the current version is 50, and then you shutdown IoTDB instance (, but at that time, the current version file is 0). 
Then, when you restore the version from the version file, you will set current version as 0. 
Is that OK?

Comment (jixuan1989) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java]: Well, I understood now..  it is correct.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    oldFile.renameTo(newFile);
+    prevVersion = currVersion;
+  }
+
+  private void restore() throws IOException {
+    File directory = new File(directoryPath);
+    File[] versionFiles = directory.listFiles((dir, name) -> name.startsWith(FILE_PREFIX));
+    File versionFile = null;
+    if (versionFiles != null && versionFiles.length > 0) {
+      Arrays.sort(versionFiles, Comparator.comparing(File::getName));
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java]: Suppose there are two version files: 900 and 1000. Then Comparator.comparing(File.getName) is incorrect.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/bufferwrite/BufferWriteProcessor.java [hunk 3]
```diff
@@ -500,4 +506,20 @@ public void setNewProcessor(boolean isNewProcessor) {
...
+   * @param deviceId the deviceId of the timeseries to be deleted.
+   * @param measurementId the measurementId of the timeseries to be deleted.
+   * @param timestamp the upper-bound of deletion time.
+   */
+  public void delete(String deviceId, String measurementId, long timestamp) {
+    workMemTable.delele(deviceId, measurementId, timestamp);
+    if (isFlush) {
+      // flushing MemTable cannot be directly modified since another thread is reading it
+      flushMemTable = flushMemTable.copy();
+      flushMemTable.delele(deviceId, measurementId, timestamp);
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/bufferwrite/BufferWriteProcessor.java]: Hm... so is this operation is just for guaranteeing `queryBufferWriteData()` ? 
Actually, `flushMemTable = flushMemTable.copy();` can not affect the flushing memtable... so you have to add the Deletion operation in the modification file at the same time.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java [hunk 4]
```diff
@@ -1902,14 +1924,43 @@ private FileNodeProcessorStore readStoreFromDisk() throws FileNodeProcessorExcep
...
+      // TODO: how to avoid partial deletion?
+      long version = versionController.nextVersion();
+
+      String fullPath = deviceId +
+              IoTDBConstant.PATH_SEPARATOR + measurementId;
+      Deletion deletion = new Deletion(fullPath, version, timestamp);
+      if (currentIntervalFileNode != null && currentIntervalFileNode.containsDevice(deviceId)) {
+        currentIntervalFileNode.getModFile().write(deletion);
+      }
+      for (IntervalFileNode fileNode : newFileNodes) {
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: write Deletion on each TsFile? (how about checking whether the startTime of the device in a TsFile is less than the given timestamp?)

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java [hunk 5]
```diff
@@ -263,6 +269,11 @@ public FileNodeProcessor(String fileNodeDirPath, String processorName)
       registStatMetadata();
       statMonitor.registStatistics(statStorageDeltaName, this);
     }
+    try {
+      versionController = new SimpleFileVersionController(fileNodeDirPath);
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: How about create a VersionControllerFactory. 
Latter we can modify the factory to switch the implementation of the versionController, otherwith SysTimestampVersionController is meaningless..

File: iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 6]
```diff
@@ -94,4 +101,58 @@ public TimeValuePairSorter query(String deviceId, String measurement, TSDataType
     return memTableMap.get(deviceId).get(measurement);
   }
 
+  @Override
+  public void delele(String deviceId, String measurementId, long timestamp) {
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: delele -> delete :D

File: iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 7]
```diff
@@ -94,4 +101,58 @@ public TimeValuePairSorter query(String deviceId, String measurement, TSDataType
...
+    Map<String, IWritableMemChunk> deviceMap = memTableMap.get(deviceId);
+    if (deviceMap != null) {
+      IWritableMemChunk chunk = deviceMap.get(measurementId);
+      IWritableMemChunk newChunk = filterChunk(chunk, timestamp);
+      if (newChunk != null) {
+        deviceMap.put(measurementId, newChunk);
+      }
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: so, how to guarantee the concurrency control.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 8]
```diff
@@ -94,4 +101,58 @@ public TimeValuePairSorter query(String deviceId, String measurement, TSDataType
...
+
+  /**
+   * If chunk contains data with timestamp less than 'timestamp', create a copy and delete all
+   * those data. Otherwise return null.
+   * @param chunk the source chunk.
+   * @param timestamp the upper-bound of deletion time.
+   * @return A reduced copy of chunk if chunk contains data with timestamp less than 'timestamp',
+   * of null.
+   */
+  private IWritableMemChunk filterChunk(IWritableMemChunk chunk, long timestamp) {
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: how about using the concept of `Page`?
Suppose a Chunk consists of page1, page2, page3,...., page n, and you will delete page 1, 2 and a part of page 3.
Then, just re-compress page 3 (and as a result, this page will have just a few points), and keep page 4~n into the new Chunk.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/TsFileWriter.java [hunk 9]
```diff
@@ -258,7 +260,7 @@ private boolean flushAllChunkGroups() throws IOException {
               chunkGroupFooter.getDataSize(), fileWriter.getPos() - pos));
         }
 
-        fileWriter.endChunkGroup(chunkGroupFooter);
+        fileWriter.endChunkGroup(chunkGroupFooter, version++);
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/TsFileWriter.java]: Suppose we have many TsFiles, and they are generated by different programs.. 
When we put all of them into an IoTDB instance (e.g., using Postback module), maybe we need a version repairing tool .
(There is no need to add the function into this PR, but if so, we need record an issue at least.)

File: iotdb/src/test/java/org/apache/iotdb/db/engine/modification/DeletionFileNodeTest.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+  @Before
+  public void setup() throws MetadataArgsErrorException,
+          PathErrorException, IOException, FileNodeManagerException {
+    MManager.getInstance().setStorageLevelToMTree(processorName);
+    for (int i = 0; i < 10; i++) {
+      MManager.getInstance().addPathToMTree(processorName + "." + measurements[i], dataType,
+              encoding, args);
+      FileNodeManager.getInstance().addTimeSeries(new Path(processorName, measurements[i]), dataType,
+              encoding, args);
```
[the remark this answers is not in the document]
Comment (jt2594838) [reply iotdb/src/test/java/org/apache/iotdb/db/engine/modification/DeletionFileNodeTest.java]: fixed

File: iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java (hunk 2 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/version/SimpleFileVersionController.java]: Replaced with scan-and-find-maximum.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/bufferwrite/BufferWriteProcessor.java (hunk 3 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/bufferwrite/BufferWriteProcessor.java]: Exactly.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java (hunk 4 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: Added.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java (hunk 5 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: Not quite necessary, SysTimestampVersionController is for test.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java (hunk 6 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: fixed

File: iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java (hunk 8 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: Good idea, but unfortunately IWritableMemChunk does not provide such interfaces. I would suggest you issue this and improve it later.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java (hunk 7 above)
Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java]: The lock of a FileNode must be acquired at the beginning of deletion.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/TsFileWriter.java (hunk 9 above)
Comment (jt2594838) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/TsFileWriter.java]: It is not a big issue. Stand-alone TsFile does not support deletion. After importing all the TsFiles, any new deletion takes effect on all of them. So the only thing we need to do is to make the version of the deletion greater than those of all existing TsFiles, which is straight-forward.

**Pull request closed without merging**